platform/view/services/db/driver/sql/common/pagination.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package common
+
+import (
+	"fmt"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/common/driver"
+	"github.com/pkg/errors"
+)
+
+type Cursor string
+
+type NoPagination struct {
+	pageInRange bool
+}
+
+func NewNoPagination() *NoPagination {
+	return &NoPagination{pageInRange: true}
+}
+
+func (p *NoPagination) Prev() (driver.Pagination, error) {
+	return &NoPagination{
+		pageInRange: false,
+	}, nil
+}
+
+func (p *NoPagination) Next() (driver.Pagination, error) {
+	return &NoPagination{
+		pageInRange: false,
+	}, nil
+}
+func (p *NoPagination) First() (driver.Pagination, error) {
+	return &NoPagination{
+		pageInRange: true,
+	}, nil
+}
+
+type OffsetPagination struct {
+	offset      int
+	pageSize    int
+	pageInRange bool
+}
+
+func NewOffsetPagination(offset int, pageSize int) (*OffsetPagination, error) {
+	if offset < 0 {
+		return nil, fmt.Errorf("offset shoud be grater than zero. Offset: %d", offset)
+	}
+	if pageSize < 0 {
+		return nil, fmt.Errorf("page size shoud be grater than zero. pageSize: %d", pageSize)
+	}
+	return &OffsetPagination{offset: offset, pageSize: pageSize, pageInRange: true}, nil
+}
+
+func (p *OffsetPagination) GoToOffset(offset int) (driver.Pagination, error) {
+	if offset < 0 {
+		return &OffsetPagination{
+			offset:      offset,
+			pageSize:    p.pageSize,
+			pageInRange: false,
+		}, nil
+	}
+	return &OffsetPagination{
+		offset:      offset,
+		pageSize:    p.pageSize,
+		pageInRange: true,
+	}, nil
+}
+
+func (p *OffsetPagination) GoToPage(pageNum int) (driver.Pagination, error) {
+	return p.GoToOffset(pageNum * p.pageSize)
+}
+
+func (p *OffsetPagination) GoForward(numOfpages int) (driver.Pagination, error) {
+	return p.GoToOffset(p.offset + (numOfpages * p.pageSize))
+}
+
+func (p *OffsetPagination) GoBack(numOfpages int) (driver.Pagination, error) {
+	return (p.GoForward(-1 * numOfpages))
+}
+
+func (p *OffsetPagination) Prev() (driver.Pagination, error)  { return p.GoBack(1) }
+func (p *OffsetPagination) Next() (driver.Pagination, error)  { return p.GoForward(1) }
+func (p *OffsetPagination) First() (driver.Pagination, error) { return p.GoToPage(0) }
+
+func NewPaginationInterpreter() *paginationInterpreter {
+	return &paginationInterpreter{}
+}
+
+type PaginationInterpreter interface {
+	Interpret(p driver.Pagination) (string, error)
+}
+
+type paginationInterpreter struct{}
+
+func (i *paginationInterpreter) Interpret(p driver.Pagination) (string, error) {
+	switch pagination := p.(type) {
+	case *NoPagination:
+		if pagination.pageInRange {
```

Comment:
keeping a `pageInRange` variable for each struct makes it a bit more cumbersome to maintain.

You can have a different type `NoResults` or `ErroredPagination` or `OutOfRangePagination` or something like that and return this every time. It seems that the other parameters don't matter if we are in this case and the behavior of the interpreter is the same independently of the pagination type.